lib/from_honeybee/geometry/aperture.rb

```diff
@@ -59,39 +59,80 @@ def to_openstudio(openstudio_model)
         os_vertices << OpenStudio::Point3d.new(vertex[0], vertex[1], vertex[2])
       end
       reordered_vertices = OpenStudio.reorderULC(os_vertices)
-
-      os_subsurface = OpenStudio::Model::SubSurface.new(reordered_vertices, openstudio_model)
-      os_subsurface.setName(@hash[:identifier])
-
-      # assign the construction if it exists
-      if @hash[:properties][:energy][:construction]
-        construction_identifier = @hash[:properties][:energy][:construction]
-        construction = openstudio_model.getConstructionByName(construction_identifier)
-        unless construction.empty?
-          os_construction = construction.get
-          os_subsurface.setConstruction(os_construction)
-        end
-      end
       
-      # assign the bondary condition object if it's a Surface
-      if @hash[:boundary_condition][:type] == 'Surface'
-        # get adjacent sub surface by identifier from openstudio model
-        adj_srf_identifier = @hash[:boundary_condition][:boundary_condition_objects][0]
-        sub_srf_ref = openstudio_model.getSubSurfaceByName(adj_srf_identifier)
-        unless sub_srf_ref.empty?
-          sub_srf = sub_srf_ref.get
-          os_subsurface.setAdjacentSubSurface(sub_srf)
+      # triangulate subsurface if neccesary
+      triangulated = false
+      final_vertices_list = []
+      if reordered_vertices.size > 4
+        
+        # transform to face coordinates
+        t = OpenStudio::Transformation::alignFace(reordered_vertices)
+        tInv = t.inverse
+        face_vertices = OpenStudio::reverse(tInv*reordered_vertices)
+        
+        # no holes in the subsurface
+        holes = OpenStudio::Point3dVectorVector.new
+        
+        # triangulate surface
+        triangles = OpenStudio::computeTriangulation(face_vertices, holes)
+        if triangles.empty?
+          raise "Failed to triangulate aperture #{@hash[:identifier]} with #{reordered_vertices.size} vertices"
```

Comment:
@chriswmackey the biggest question i had was how you want to handle errors in triangulation?

Reply:
Is there a list of known cases that fail here?  I understand that there are always possible situations of poor input (like non-planar shapes or self-intersecting bow-tie shapes). But if it's anything like the ear clipping method that we currently use in Python, this shouldn't be failing for any of the cases that we permit from the plugin side. So I think raising an exception like this is fine.

Reply:
This triangulation should be pretty robust but there is always some crazy edge case that will come up.

Reply:
Understood. I think this is fine, then.